File: May-LeetCoding-Challenge/09-Valid-Perfect-Square/Valid-Perfect-Square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function(num) {
    if(num==1){return true;}
    var i = 1;
    var j = parseInt(num/2);
    while(i<j){
        let mid = parseInt((i+j)/2);
        let square = mid*mid;
        if(i*i==num || j*j==num||square==num){
            return true;
        }
        else if(square<num){
            i = mid+1;
        }
        else{
            j = mid-1;
        }
    }
    return false;
};

File: May-LeetCoding-Challenge/09-Valid-Perfect-Square/Valid-Perfect-Square.scala
object Solution {
  def isPerfectSquare(num: Int): Boolean = isPerfectSquareWithRange(num, (0, 46341))

  def isPerfectSquareWithRange(num: Int, range: (Int, Int)): Boolean = {
    if (range._2 - range._1 < 2) false
    else {
      val mid = (range._1 + range._2) / 2
      val midSqr = mid * mid
      if (midSqr == num) true
      else isPerfectSquareWithRange(num, if (midSqr > num) (range._1, mid) else (mid, range._2))
    }
  }
}
